Created API endpoint for creating Coupon orders.

vkaracic opened the pull request:
This PR contains the API endpoint for coupon products, creating coupons, adding them to the basket, handling the order, fulfilling the order, with accompanying tests. (the test for the fulfillment module is not yet finished).

@clintonb can you please take a look at the code? Also note that #450 is a prerequisite for this so there is code that calls some models and methods from that PR, Travis will probably fail because of that too. But for now let's just fix any code issues that there might be. Thanks!

@mjfrey @marjev @dsego

clintonb commented:
@rlucioni @jimabramson can one of you take a look at this as well, please?

clintonb commented on ecommerce/extensions/api/serializers.py:
```diff
@@ -332,3 +333,27 @@ def get_products(self, obj):
+    vouchers = serializers.SerializerMethodField()
+
+    def get_price(self, obj):
+        stock_record = StockRecord.objects.get(product=obj)
```
See `ProductSerializer.get_price()`.

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -0,0 +1,236 @@
+
+    def test_list_coupons(self):
+        """Test coupon API endpoint list."""
+        response = self.client.get('/api/v2/coupons/')
```
Use `reverse`

clintonb replied:
API tests belong in the api app.

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -0,0 +1,236 @@
+
+        # If there are no products it returns an empty response.
+        Product.objects.all().delete()
+        response = self.client.get('/api/v2/coupons/')
```
Use `reverse`.

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -0,0 +1,236 @@
+        self.catalog = Catalog.objects.create(partner=self.partner)
+        self.product_class, __ = ProductClass.objects.get_or_create(name='Coupon')
+
+        self.data = {
```
This does not need to be an instance variable.

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -0,0 +1,236 @@
+            'voucher_type': Voucher.SINGLE_USE
+        }
+
+        self.coupon = CouponOrderCreateView().create_coupon_product(
```
This isn't needed by all tests. Move to a method, and call only when needed.

clintonb replied:
Also, use a factory.

clintonb commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -0,0 +1,263 @@
+StockRecord = get_model('partner', 'StockRecord')
+
+
+class CouponOrderCreateView(generics.CreateAPIView, EdxOrderPlacementMixin, NonDestroyableModelViewSet):
```
Combining `NonDestroyableModelViewSet` and `CreateAPIView` is odd. Why are you doing this?

clintonb commented on ecommerce/extensions/api/v2/urls.py:
```diff
@@ -21,6 +23,14 @@
+]
+
+VOUCHER_URLS = [
+    url(r'^$', vouchers_views.VoucherViewSet.as_view({'get': 'list'}), name='list'),
```
If you only need one access method, why bother with a viewset?

clintonb commented on ecommerce/extensions/api/v2/urls.py:
```diff
@@ -21,6 +23,14 @@
 ]
 
+COUPON_URLS = [
+    url(r'^$', coupons_views.CouponOrderCreateView.as_view({'get': 'list'}), name='create'),
```
If you only need one access method, why bother with a viewset?

clintonb commented on ecommerce/extensions/catalogue/utils.py:
```diff
@@ -1,22 +1,73 @@
+    return catalog
+
+
+def generate_upc(partner, title, catalog):
```
We don't actually use UPCs. Also, UPCs follow a standard: https://en.wikipedia.org/wiki/Universal_Product_Code. This should be removed.

vkaracic replied:
Unfortunately there is no other way to differentiate a coupon (except for creating and deleting objects in the same method which wouldn't be ideal). To be precise, in the time of creating a coupon the only elements for creating it are ``title`` (which is not unique) and ``product_class`` (a constant) - none of those can be used as parameters for a ``get_or_create`` function.

mjfrey replied:
So everyone is in agreement.  The goal here is to create a unique product and re-use / update an existing one if the same parameters are used in the Coupon creation form.

clintonb replied:
If that's the use case, this works. Just don't call it UPC.

vkaracic replied:
It has to be UPC, because that's the only unique attribute that we can use from Oscar's Product model: https://github.com/django-oscar/django-oscar/blob/master/src/oscar/apps/catalogue/abstract_models.py#L267

I guess I could change the algorithm of creating it a bit. Or we'll need to adjust the Product model which we're trying hard to avoid.

clintonb replied:
UPC is not the only unique attribute. We use a code/slug field for seats.
See Course.create_or_update_seat().

-- 
Clinton Blackburn

vkaracic replied:
Okay, I've overlooked the ``slug`` attribute in the Product model, sorry.

clintonb commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -0,0 +1,263 @@
+            partner=data['partner']
+        )
+        product_class = ProductClass.objects.get(slug='coupon')
+        coupon_product, created = Product.objects.get_or_create(
```
Use some combination of title, catalog, and partner to differentiate the products you are creating.

clintonb commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -0,0 +1,263 @@
+                partner=data['partner'],
+                partner_sku=sku
+            )
+            stock_record.price_currency = 'USD'
```
Move these to `create` and avoid the extra DB call.

clintonb commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -0,0 +1,263 @@
+            catalog=data['catalog'],
+        )
+
+        if created:
```
These two cases can be combined:

```python
stock_record, _created = StockRecord.objects.get_or_create(product=coupon_product, partner=data['partner'], partner_sku=sku)
stock_record.price_currency = 'USD'
stock_record.price_excl_tax = price
stock_record.save()
```

clintonb commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -0,0 +1,263 @@
+
+        # Product validation.
+        try:
+            coupon_product.clean()
```
Shouldn't this be validated before we go through the effort of creating stock records?

clintonb commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -0,0 +1,263 @@
+        basket.add_product(product)
+        logger.info(
+            'Added product with SKU [%s] to basket [%d]',
+            product.stockrecords.first().partner_sku, basket.id
```
That's now how you get the SKU. Use `product.stockrecords.filter(partner=get_partner_for_site(site)).partner_sku`.

vkaracic replied:
``get_partner_for_site`` expects the ``request`` as an argument. Can I just send the ``partner`` and:
``product.stockrecords.filter(partner=partner).partner_sku`` instead of the whole request?

clintonb replied:
That works.

clintonb commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -0,0 +1,263 @@
+
+        # Invoice payment processor invocation.
+        payment_processor = InvoicePayment
+        payment_processor().handle_processor_response(response={}, basket=basket)
```
Payment processing should occur before order placement.

vkaracic commented:
I don't quite understand the test error:
``OperationalError: no such table: catalogue_productclass``
ProductClass is Oscar's model.

vkaracic commented:
@clintonb I believe I have addressed all your comments, I'll work on fixing the coverage, in the meantime can you take a second look?

clintonb commented on ecommerce/extensions/api/serializers.py:
```diff
@@ -332,3 +339,34 @@ def get_products(self, obj):
+    price = serializers.SerializerMethodField()
+    vouchers = serializers.SerializerMethodField()
+
+    def get_price(self, product):
```
Move this, `get_info`, and the `price` field to a mixin shared with `ProductSerializer`.

vkaracic replied:
I've created a new mixin.

clintonb commented on ecommerce/extensions/api/v2/views/vouchers.py:
```diff
@@ -0,0 +1,15 @@
+class VoucherViewSet(viewsets.ReadOnlyModelViewSet):
+    queryset = Voucher.objects.all()
+    serializer_class = serializers.VoucherSerializer
+    permission_classes = (IsAuthenticated, IsAdminUser,)
```
What's the use case of this API?

vkaracic replied:
At some point we used it, but we've moved the listing to the coupons API. It's dead code, I'll remove it.

vkaracic commented:
@clintonb quality satisfied, and your last comments addressed, can you please take another look?

clintonb commented on ecommerce/extensions/api/mixins.py:
```diff
@@ -0,0 +1,19 @@
+Selector = get_class('partner.strategy', 'Selector')
+
+
+class ProductInfoMixin(serializers.ModelSerializer):
```
1. This can/should reside in `serializers.py`.
2. Make it clear that this is intended for mixing with serializers, and avoid weird inheritance issues: `class ProductInfoSerializerMixin(object):`.

vkaracic replied:
Actually when I inherit from ``object`` I get ``ImproperlyConfigured`` error:
```
Field name `price` is not valid for model `Product`.
```

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -0,0 +1,284 @@
+        }
+        request = RequestFactory()
+        request.data = data
+        request.site = site
```
You can use `self.site`.

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -0,0 +1,284 @@
+        request.data = data
+        request.site = site
+
+        response = CouponOrderCreateView().create(request)
```
Call the API via HTTP: `self.client.get(...)`.

clintonb replied:
I see what you're doing. You're actually separating unit/functional tests. That's fine. We generally start at the functional level.

vkaracic replied:
I'm not trying to call the API, but to test the ``create`` method (which expects a request).

clintonb commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -0,0 +1,240 @@
+StockRecord = get_model('partner', 'StockRecord')
+
+
+class CouponOrderCreateView(EdxOrderPlacementMixin, NonDestroyableModelViewSet):
```
If this is just a creation view, you don't need a `ViewSet`.

clintonb commented on ecommerce/extensions/catalogue/migrations/0013_coupon_product_class.py:
```diff
@@ -24,7 +24,7 @@ def create_product_class(apps, schema_editor):
         code='coupon_vouchers',
         type='entity',
-        required=False
+        required=True
```
You cannot modify migrations this way. This migration has already been run in production. Make a new migration with your changes.

clintonb commented on ecommerce/extensions/catalogue/tests/test_utils.py:
```diff
@@ -14,9 +33,79 @@ def test_generate_sku_for_course_seat(self):
         certificate_type = 'honor'
         product = course.create_or_update_seat(certificate_type, False, 0, self.partner)
 
-        _hash = u'{} {} {} {} {}'.format(certificate_type, course_id, 'False', '', self.partner.id)
```
Why the change?

vkaracic replied:
When there is ``from __future__ import unicode_literals`` is it needed anymore to explicitly add ``u`` in front of strings?

clintonb replied:
Okay.

clintonb commented on ecommerce/extensions/fulfillment/modules.py:
```diff
@@ -281,3 +281,52 @@ def revoke_line(self, line):
+        return order, lines
+
+    def revoke_line(self, line):  # pragma: no cover
+        raise NotImplementedError("Revoke method not implemented!")
```
How will this be implemented?

vkaracic replied:
@mfrey

On Fri, Dec 11, 2015 at 2:33 PM, Clinton Blackburn <notifications@github.com
> wrote:

> In ecommerce/extensions/fulfillment/modules.py
> <https://github.com/edx/ecommerce/pull/453#discussion_r47354769>:
>
> > +            order (Order): The Order associated with the lines to be fulfilled.
> > +            lines (List of Lines): Order Lines, associated with purchased products in an Order. These should only
> > +                be 'Coupon' products.
> > +        Returns:
> > +            The original set of lines, with new statuses set based on the success or failure of fulfillment.
> > +        """
> > +        logger.info("Attempting to fulfill 'Coupon' product types for order [%s]", order.number)
> > +
> > +        for line in lines:
> > +            line.set_status(LINE.COMPLETE)
> > +
> > +        logger.info("Finished fulfilling 'Coupon' product types for order [%s]", order.number)
> > +        return order, lines
> > +
> > +    def revoke_line(self, line):  # pragma: no cover
> > +        raise NotImplementedError("Revoke method not implemented!")
>
> How will this be implemented?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/ecommerce/pull/453/files#r47354769>.
>



-- 
Vedran Karačić
Software developer with ExtensionEngine

Mail:     vedran@edx.org
             vkaracic@extensionengine.com
Github: vkaracic <https://github.com/vkaracic>
Skype:  karacicvedran

mjfrey replied:
I don't think this has been decided yet.  I have had a discussion with product regarding refunds of Coupons, but no decision has been made.  For example what if 5 of 10 Vouchers have been redeemed and then we want to refund. Not sure how we do that.

mfrey replied:
@vkaracic could you please remove me from your comment? i'm obviously not @mjfrey and i don't want to receive notifications about this pull request. thanks!

clintonb replied:
@mjfrey you can simply unsubscribe from the PR.

vkaracic replied:
@mfrey sorry, was a typo :)

dsego replied:
:joy:

marjev replied:
I think you don't need to re-implement this method. It should be inherited from `BaseFulfillmentModule`.

vkaracic replied:
I thought so too, but once I remove it I get this error:
```
2015-12-14 03:43:38,305 ERROR 6889 [ecommerce.extensions.fulfillment.api] api.py:67 - An unexpected error occurred while fulfilling order [EDX-100001].
Traceback (most recent call last):
  File "/home/vedran/Development/edX/otto/ecommerce/ecommerce/extensions/fulfillment/api.py", line 54, in fulfill_order
    module = module_class()
TypeError: Can't instantiate abstract class CouponFulfillmentModule with abstract methods revoke_line
```

clintonb commented:
@vkaracic minor issues identified. Let me know when they are addressed, and I'll review a final time.

vkaracic commented:
@clintonb ready for final review. Note that I have removed the coupons serializer because the coupons api endpoint won't list coupons, I've moved that to the products api endpoint where it belongs, therefor there won't be a need to the mixin class.

clintonb commented on ecommerce/extensions/catalogue/migrations/0014_alter_couponvouchers_attribute.py:
```diff
@@ -0,0 +1,25 @@
+        ('catalogue', '0013_coupon_product_class')
+    ]
+    operations = [
+        migrations.RunPython(alter_couponvouchers_attribute)
```
Reverse migration needed.

clintonb commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -0,0 +1,237 @@
+            logger.exception('Failed to create vouchers for [%s] coupon.', coupon_product.title)
+            raise IntegrityError(ex)  # pylint: disable=nonstandard-exception
+
+        coupon_vouchers = CouponVouchers.objects.get(coupon=coupon_product)
```
@mjfrey remind me why we are creating vouchers here instead of during fulfillment.

mjfrey replied:
@clintonb  We discussed this last week.  The first implementation was during fulfillment, but you suggested moving it to before.  I can not actually remember the reason why.  I think we discussed having a real invoice processor so we wanted the vouchers created before and not in fulfillment.

mjfrey replied:
I think we also discussed the fact that we wanted the vouchers to part of the order as lines so we could reference them.  If we did them in fulfillment they would not be part of the line(s)

clintonb replied:
Okay.

clintonb replied:
Please add a comment explaining why we are creating the order immediately.

mjfrey commented:
@clintonb can we get a last look

clintonb commented:
Rebase on master to fix the migration bug.

edx-webhook commented:
Thanks for the pull request, @vkaracic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fecommerce&number=453

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -0,0 +1,278 @@
+        self.assertEqual(Basket.objects.first().status, 'Submitted')
+
+
+class CouponOrderCreateViewFunctionalTest(TestCase):
```
You need to test authentication/authorization.

clintonb commented on ecommerce/extensions/catalogue/migrations/0014_alter_couponvouchers_attribute.py:
```diff
@@ -0,0 +1,32 @@
+ProductAttribute = get_model("catalogue", "ProductAttribute")
+
+
+def alter_couponvouchers_attribute(apps, shema_editor):
```
schema_editor

clintonb commented on ecommerce/extensions/catalogue/migrations/0014_alter_couponvouchers_attribute.py:
```diff
@@ -0,0 +1,32 @@
+    coupon_vouchers.save()
+
+
+def reverse_migration(apps, shema_editor):
```
schema_editor

clintonb commented on ecommerce/extensions/catalogue/utils.py:
```diff
@@ -1,22 +1,74 @@
+        catalog = kwargs.get('catalog', '')
+        _hash = ' '.join((
+            unicode(product.id),
+            unicode(catalog.name),
```
The name can change. Use `catalog.id`.

clintonb commented on ecommerce/extensions/fulfillment/modules.py:
```diff
@@ -281,3 +281,52 @@ def revoke_line(self, line):
         return False
+
+
+class CouponFulfillmentModule(BaseFulfillmentModule):
```
This needs to be tested.

vkaracic commented:
@clintonb I have updated the PR and addressed your latests comments. A couple of things to note, I have added the viewset again because we will be listing coupon products in the /coupons API endpoint (https://github.com/edx/ecommerce/pull/460#discussion_r47595671). The other thing is that I removed this dispatch method: https://github.com/vkaracic/ecommerce/commit/d31ef2359b99bae7fca79617a982e44861b1eb78#diff-c9a451f0afeebe0481e4cee1c16b7c94L44 because it gave out 
``TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.``
errors in the authentication/authorization tests. Without it all the tests pass, so I don't think that we need it, but I would like to hear your opinion about it.

clintonb commented on ecommerce/extensions/api/serializers.py:
```diff
@@ -26,11 +26,28 @@
 PRODUCT_DETAIL_VIEW = 'api:v2:product-detail'
 
 
+class ProductInfoSerializerMixin(serializers.ModelSerializer):
```
This should inherit for `object`, not `ModelSerializer`. I'm surprised there haven't been MRO errors.

vkaracic replied:
I get MRO errors when inheriting from ``serializers.Serializer``, when inheriting from object: https://github.com/edx/ecommerce/pull/453#discussion_r47650815

clintonb replied:
Leave as-is. I don't have a quick fix and, ultimately the price field needs to go away in lieu of stock records.

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -0,0 +1,295 @@
+        self.assertEqual(Order.objects.first().lines.count(), 1)
+        self.assertEqual(Order.objects.first().lines.first().product.title, 'Test coupon')
+
+    def test_authorization_required(self):
```
`test_authentication_required`

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -0,0 +1,295 @@
+        response = self.client.get(COUPONS_LINK)
+        self.assertEqual(response.status_code, 401)
+
+    def test_authentication_required(self):
```
`test_authorization_required`

clintonb commented:
@vkaracic 
1. Rather than add another endpoint exposing products, add filtering to the existing one.
2. Did you investigate the error that occurred in the atomic block before removing it? The atomicity code has worked in production for few months. Please investigate the error to make sure removing that code is the best solution.

clintonb commented:
I'm seeing `catalog.name` used in many places. Names can change. IDs cannot. Consider using `id` instead of `name`, especially in places where we are generating codes/slugs that will be used for uniqueness or matching.

mjfrey commented:
@clintonb we did have this under the /products endpoint but you suggested moving it to its own since there are a lot of products.  So lets pick one way and stick to it.

clintonb commented:
Did you have general filtering, or a separate /api/v2/products/coupons/
endpoint? I am proposing generic filtering by product class (e.g.
/api/v2/products/?product_class=coupon). See
http://www.django-rest-framework.org/api-guide/filtering/#generic-filtering.

clintonb commented on ecommerce/extensions/fulfillment/modules.py:
```diff
@@ -281,3 +281,52 @@ def revoke_line(self, line):
+        logger.info("Finished fulfilling 'Coupon' product types for order [%s]", order.number)
+        return order, lines
+
+    def revoke_line(self, line):  # pragma: no cover
```
Add a test, making sure this always raises an error. If someone comes along and changes this method, we want to make sure the changes are tested.

clintonb commented on ecommerce/extensions/api/serializers.py:
```diff
@@ -49,6 +50,11 @@ def get_name(self, instance):
         return instance.attribute.name
 
     def get_value(self, obj):
+        if obj.attribute.name == 'Coupon vouchers':
```
This needs to be tested.

clintonb commented on ecommerce/extensions/api/serializers.py:
```diff
@@ -332,3 +349,23 @@ def get_products(self, obj):
         )
+
+
+class VoucherSerializer(serializers.ModelSerializer):
```
This needs to be tested.

clintonb commented:
:+1: (once tests and voucher comment added)

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_products.py:
```diff
@@ -28,6 +32,28 @@ def setUp(self):
 
+    def create_coupon(self):
+        """Helper method for creating a coupon."""
+        catalog = Catalog.objects.create(partner=self.partner)
```
Create a factory for this.

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_products.py:
```diff
@@ -28,6 +32,28 @@ def setUp(self):
+            'voucher_type': Voucher.SINGLE_USE
+        }
+
+        coupon = CouponOrderCreateView().create_coupon_product(
```
Create a factory.

clintonb commented on ecommerce/extensions/api/v2/views/products.py:
```diff
@@ -13,4 +14,9 @@
 class ProductViewSet(NestedViewSetMixin, NonDestroyableModelViewSet):
     queryset = Product.objects.all()
     serializer_class = serializers.ProductSerializer
+    # URL with filter example:
```
This belongs in a docstring on the class.

clintonb commented on ecommerce/extensions/api/v2/views/products.py:
```diff
@@ -13,4 +14,9 @@
     serializer_class = serializers.ProductSerializer
+    # URL with filter example:
+    # /api/v2/products/?product_class__name=Coupon
+    # Note, it's case-sensitive
```
Use a FilterSet to make this case-insensitive:
* http://www.django-rest-framework.org/api-guide/filtering/#specifying-a-filterset
* https://django-filter.readthedocs.org/en/latest/usage.html

vkaracic commented:
@clintonb I believe I have addressed all your comments, and I squashed the commits.

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_products.py:
```diff
@@ -120,3 +124,47 @@ def test_no_partner_product(self):
+        self.assertEqual(response_data['attribute_values'][0]['name'], 'Coupon vouchers')
+        self.assertEqual(len(response_data['attribute_values'][0]['value']), 5)
+
+    def test_coupon_voucher_serializer(self):
```
Ideally serializers should be tested separately of the view. Besides giving us definite test coverage of the serializers, this testing also allows us to confidently use the serializer to generate expected output.

Here is an example in the edx/course-discovery:
* Serializer tests: https://github.com/edx/course-discovery/blob/clintonb/courses/course_discovery/apps/api/tests/test_serializers.py
* Tests utilizing serializers: https://github.com/edx/course-discovery/blob/clintonb/courses/course_discovery/apps/api/v1/tests/test_views.py#L97-L104

clintonb commented:
@vkaracic looks good. I'd prefer to see explicit serializer tests. Your choice.

vkaracic commented:
@clintonb I couldn't find any existing serializer tests so I though you guys test it that way. I could do it as a separate task, so feel free to create a task/story on JIRA and assign me. But for now we need to get this merged since there's a lot of work pending this merge.
